scheds/rust/scx_bpfland/src/bpf/main.bpf.c

```diff
@@ -0,0 +1,490 @@
+	 *
+	 * In general, it is more convenient to rely on the vruntime-based
+	 * scheduling (via bpfland_enqueue()) to guarantee a better level of
+	 * responsiveness, but in certain cases we may get better performance
+	 * dispatching tasks directly here (depending on the particular
+	 * architecture), relying on the built-in idle selection logic.
+	 */
+	if (!builtin_idle)
+		return prev_cpu;
+
+	/*
+	 * Find another idle CPU and potentially migrate the task.
+	 */
+	cpu = scx_bpf_select_cpu_dfl(p, prev_cpu, wake_flags, &is_idle);
+	if (is_idle) {
+		dispatch_direct_cpu(p, cpu);
+		return cpu;
+	}
+
+	return prev_cpu;
+}
+
+void BPF_STRUCT_OPS(bpfland_enqueue, struct task_struct *p, u64 enq_flags)
+{
+	/*
+	 * Always dispatch per-CPU kthreads immediately.
+	 *
+	 * This allows to prioritize critical kernel threads that may
+	 * potentially slow down the entire system if they are blocked for too
+	 * long (i.e., ksoftirqd/N, rcuop/N, etc.).
+	 */
+	if (is_kthread(p) && p->nr_cpus_allowed == 1) {
+		dispatch_kthread(p, enq_flags);
+		return;
+	}
+
+	/*
+	 * Dispatch all the other tasks that were not dispatched directly in
+	 * select_cpu().
+	 */
+	dispatch_task(p, enq_flags);
+}
+
+void BPF_STRUCT_OPS(bpfland_dispatch, s32 cpu, struct task_struct *prev)
+{
+	/*
+	 * First consume directly dispatched tasks, so that they can
+	 * immediately use the CPU assigned in select_cpu().
+	 */
+	if (scx_bpf_consume(cpu_to_dsq(cpu)))
+		return;
+
+	/*
+	 * Then always consume interactive tasks before regular tasks.
+	 *
+	 * This is fine and we shouldn't have starvation, because interactive
+	 * tasks are classified by their amount of voluntary context switches,
+	 * so they should naturally release the CPU quickly and give a chance
+	 * to the regular tasks to run.
+	 */
+	if (scx_bpf_consume(PRIO_DSQ))
+		return;
+
+	/*
+	 * Lastly, consume regular tasks from the shared DSQ.
+	 */
+	scx_bpf_consume(SHARED_DSQ);
+}
+
+void BPF_STRUCT_OPS(bpfland_running, struct task_struct *p)
+{
+	/* Update global vruntime */
+	if (vtime_before(vtime_now, p->scx.dsq_vtime))
+		vtime_now = p->scx.dsq_vtime;
+
+	__sync_fetch_and_add(&nr_running, 1);
+}
+
+/*
+ * Update task statistics when the task is releasing the CPU (either
+ * voluntarily or because it expires its assigned time slice).
+ */
+void BPF_STRUCT_OPS(bpfland_stopping, struct task_struct *p, bool runnable)
+{
+	u64 slice = slice_ns - p->scx.slice;
+	u64 weight = p->scx.weight;
+	u64 now = bpf_ktime_get_ns();
+	struct task_ctx *tctx;
+
+	__sync_fetch_and_sub(&nr_running, 1);
+
+	tctx = lookup_task_ctx(p);
+	if (!tctx)
+		return;
+
+	/*
+	 * Classify interactive tasks based on the average amount of their
+	 * voluntary context switches and automatically boost their priority
+	 * (that will affect their vruntime).
+	 */
+	if (nvcsw_thresh && weight_boost) {
+		if (tctx->avg_nvcsw >= nvcsw_thresh)
+			tctx->is_interactive = true;
+		if (tctx->is_interactive)
+			weight *= weight_boost;
```

Comment:
Thanks for noticing it, you're absolutely right! That was basically copied from rustland, in the first implementation of bpfland I was dispatching all the tasks to the same `SHARED_DSQ` and I wasn't very happy about performance, so I tried to add a separate `PRIO_DSQ` and it was definitely better. In fact, I'm planning to apply the same change to rustland as well (I already have a patch in my local repo). With that in place the whole "priority boost" logic becomes useless. Therefore, we can definitely drop the priority boosting, I'll update this in the next push.

Side note: it's interesting to notice that `scx_rustland` has been used to prototype `scx_bpfland` and vice versa (for the `PRIO_DSQ`), so it was actually useful to re-implement the same scheduler in a different way, this could be yet another example to prove the usefulness of having easy-to-test pluggable schedulers.